juju/juju
# CMR relation remove

## Description of change

There's a few issues for CMR relations. The main one is the workers in the consuming model are stopped as soon as the relation dead event arrives, but there could still be a unit scope change event to arrive (different watchers are used). So ensure that the watchers are only stopped when all units have left scope. This is done by adding unit count to the relation-unit change event. Also, when a relation is suspended, only stop the watchers if the relation is alive, otherwise we might miss departing events.

Reduce the cmr worker restart delay to 15 seconds, as 1 minute seems too long (we want to avoid spamming when using cross controller calls hence it's not the standard 3 seconds).

We were also marking the saas status as terminated if the offering side had removed the relation, but this should only be done if the offer (or app) is removed. So add a new watcher to provide that information. Extra unit tests in various places were added also.

There's 3 commits:
- state changes
- apiserver changes
- cmr worker changes

## QA steps

```sh
juju bootstrap lxd
juju add-model database
juju deploy percona-cluster mysql --config min-cluster-size=3 -n 3
juju offer mysql:shared-db
juju switch default
juju deploy vault -n 3
juju consume database.mysql rmysql
juju relate rmysql vault
```

wait for things to settle, then 
juju remove-relation vault rmysql
The relation should be removed both consuming side and offering side.

Remove the offer.
The saas status on the consuming side should go to "terminated".

## Bug reference

https://bugs.launchpad.net/bugs/1879645
